src/Servers/Kestrel/Transport.Sockets/src/Internal/IOQueue.cs

```diff
@@ -22,35 +24,58 @@ public override void Schedule(Action<object> action, object state)
                 State = state
             };
 
+            // Enqueue prior to checking _doingWork.
             _workItems.Enqueue(work);
 
-            lock (_workSync)
+            // Fast check if already doing work
+            if (_doingWork == 0)
             {
-                if (!_doingWork)
+                // Not working, set as working, and check if it was already working (via atomic Interlocked).
+                if (Interlocked.Exchange(ref _doingWork, 1) == 0)
                 {
+                    // Wasn't working, schedule.
                     System.Threading.ThreadPool.UnsafeQueueUserWorkItem(this, preferLocal: false);
-                    _doingWork = true;
                 }
             }
         }
 
         void IThreadPoolWorkItem.Execute()
         {
+            var workItems = _workItems;
```

Comment:
Why does this matter for a readonly field? If it does matter, is there a corresponding JIT bug?

Reply:
Probably doesn't; it does for array.Length though https://github.com/dotnet/coreclr/issues/21951

Reply:
Using the field